### PR TITLE
Add _pre and _post decorators memtier.c

### PR DESCRIPTION
**Pull request by jacekgrzybek**
### Types of changes

- [x] modified configure.ac
- [ ] add add test for this for wrapped functions (malloc, calloc, etc)
- [ ] add test for this decorators

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/646)

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -194,6 +210,19 @@ fi
 
 AC_SUBST(memkind_prefix)
 
+#============================memtier_prefix===========================================
```
Comment (PatKamin) [inline configure.ac]: It shouldn't be part of this commit.

**Review by PatKamin — COMMENTED** [review]
Remove merge commits from this PR.

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 3 of 3 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/646)* status: all files reviewed, 2 unresolved discussions (waiting on @jacekgrzybek)

___
*[configure.ac, line 77 at r3](https://reviewable.io/reviews/memkind/memkind/646#-M_uMIuNAVTpUjoJoQNe:-M_uMIuNAVTpUjoJoQNf:b11gdjn) ([raw file](https://github.com/memkind/memkind/blob/c4be4422842306e05a7a313c74d0b2d13234810f/configure.ac#L77)):*
> ```M4Sugar
>   [AS_HELP_STRING([--enable-decorators-memtier], [Enable decorators])],
> [if test "x$enable_decorators_memtier" = "xyes" ; then
>   enable_decorators_memtier="1"
> ```

how this works? where is the code that is enabled with this flag? where are tests for this? Can we discus this feature online?

**Comment by jacekgrzybek** [comment]
> > Reviewed 3 of 3 files at r1, 1 of 1 files at r2.
> > _[Reviewable](https://reviewable.io/reviews/memkind/memkind/646)_ status: all files reviewed, 2 unresolved discussions (waiting on @jacekgrzybek)
> 
> _[configure.ac, line 77 at r3](https://reviewable.io/reviews/memkind/memkind/646#-M_uMIuNAVTpUjoJoQNe:-M_uMIuNAVTpUjoJoQNf:b11gdjn) ([raw file](https://github.com/memkind/memkind/blob/c4be4422842306e05a7a313c74d0b2d13234810f/configure.ac#L77)):_
> 
> > ```m4
> >   [AS_HELP_STRING([--enable-decorators-memtier], [Enable decorators])],
> > [if test "x$enable_decorators_memtier" = "xyes" ; then
> >   enable_decorators_memtier="1"
> > ```
> 
> how this works? where is the code that is enabled with this flag? where are tests for this? Can we discus this feature online?

It's code to

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/646)* status: 2 of 9 files reviewed, 3 unresolved discussions (waiting on @bratpiorka and @jacekgrzybek)

___
*[configure.ac, line 82 at r4](https://reviewable.io/reviews/memkind/memkind/646#-Mbevvmh7qegHADDLAhc:-Mbevvmh7qegHADDLAhd:b-h1susy) ([raw file](https://github.com/memkind/memkind/blob/f9ad03626c72dc91559182b2d743da90140a1622/configure.ac#L82)):*
> ```
> enable_decorators_memtier
> ```
I think we could use in a slight difference compared to MEMKIND_DECORATION_ENABLED:

- we keep the MEMTIER_DECORATION_ENABLED
- instead of defining __weak__ references we just define a functions on memkind_memtier layer

Something like:
<details><summary><i alt="summary">Quoted 9 lines of code…</i></summary>

> ```
> MEMKIND_EXPORT void *memtier_kind_malloc(memkind_t kind, size_t size) 
> {
>  void *ptr = memkind_malloc(kind, size);
>  increment_alloc_size(kind->partition, jemk_malloc_usable_size(ptr));
>   #ifdef MEMTIER_DECORATION_ENABLED
>   log_info("memtier_malloc kind:%s, (%zu) = %p", kind->name, size, ret);
>   #endif
>  return ptr;
> }
> ```
</details>


We will achieve something like:
https://github.com/memkind/memkind/blob/545d7241913d7993ea8b65afac96e3548f234a61/tiering/memtier.c#L47
with additional kind name - I think that was the main goal here.

One difference will be that the library must be configured explicitly with MEMTIER_DECORATION_ENABLED to use this feature, and MEMKIND_DEBUG is set to some variable - we could extend memkind logging mechanism to "always".

@patkamin, @jschmieg @bratpiorka let me know what you think.

**Review by PatKamin — COMMENTED** [review]
Remove merge commits from PR.

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 1 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/646#-:-M_tdkwa0VHUwQfYLX6q:b-vh6mx3)* status: 3 of 9 files reviewed, 4 unresolved discussions (waiting on @bratpiorka, @jacekgrzybek, @jschmieg, and @PatKamin)

___
*[configure.ac, line 82 at r4](https://reviewable.io/reviews/memkind/memkind/646#-Mbevvmh7qegHADDLAhc:-Mbf2rgGAy8CasPjfsn7:b-ba7jcz) ([raw file](https://github.com/memkind/memkind/blob/f9ad03626c72dc91559182b2d743da90140a1622/configure.ac#L82)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> enable_decorators_memtier
> ```
I think we could use in a slight difference compared to MEMKIND_DECORATION_ENABLED:

- we keep the MEMTIER_DECORATION_ENABLED
- instead of defining __weak__ references we just define a functions on memkind_memtier layer

Something like:

> ```
> MEMKIND_EXPORT void *memtier_kind_malloc(memkind_t kind, size_t size) 
> {
>  void *ptr = memkind_malloc(kind, size);
>  increment_alloc_size(kind->partition, jemk_malloc_usable_size(ptr));
>   #ifdef MEMTIER_DECORATION_ENABLED
>   log_info("memtier_malloc kind:%s, (%zu) = %p", kind->name, size, ret);
>   #endif
>  return ptr;
> }
> ```

We will achieve something like:
https://github.com/memkind/memkind/blob/545d7241913d7993ea8b65afac96e3548f234a61/tiering/memtier.c#L47
with additional kind name - I think that was the main goal here.

One difference will be that the library must be configured explicitly with MEMTIER_DECORATION_ENABLED to use this feature, and MEMKIND_DEBUG is set to some variable - we could extend memkind logging mechanism to "always".

@patkamin, @jschmieg @bratpiorka let me know what you think.

</blockquote></details>

Sounds good to me.
___
*[test/decorator_memtier_test, line 1 at r4](https://reviewable.io/reviews/memkind/memkind/646#-Mbf2zY-3CSDSO5yhZze:-Mbf2zY-3CSDSO5yhZzf:b-ctktgs) ([raw file](https://github.com/memkind/memkind/blob/f9ad03626c72dc91559182b2d743da90140a1622/test/decorator_memtier_test#L1)):*
> ```
> #! /bin/bash
> ```

Remove this file from the commit. Add it to .gitignore.

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 7 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/646)* status: all files reviewed, 3 unresolved discussions (waiting on @jacekgrzybek and @jschmieg)

___
*[configure.ac, line 82 at r4](https://reviewable.io/reviews/memkind/memkind/646#-Mbevvmh7qegHADDLAhc:-MbfC3mKDHolshbYktlS:b-megqkt) ([raw file](https://github.com/memkind/memkind/blob/f9ad03626c72dc91559182b2d743da90140a1622/configure.ac#L82)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Sounds good to me.
</blockquote></details>

I think logging of kind in tiering should be enabled by setting MEMKIND_MEM_TIERING_LOG_LEVEL env to debug. Otherwise we don't need this env at all and we could move all logging to memkind_memtier.c and use only MEMKIND_DEBUG flag. Also currently we log some info in memkind/tiering/memtier.c, so if we add other logging in memkind_memtier.c this log message will be obsolete and should be removed.

**Comment by michalbiesek** [comment]
Done in #680
